d2/src/main/java/com/linkedin/d2/balancer/clients/RetryClient.java

```diff
@@ -254,16 +334,200 @@ public void onError(Throwable e)
 
     private boolean isRetryException(Throwable e)
     {
-      return ExceptionUtils.indexOfType(e, RetriableRequestException.class) != -1;
+      Throwable[] throwables = ExceptionUtils.getThrowables(e);
+
+      for (Throwable throwable: throwables)
+      {
+        if (throwable instanceof RetriableRequestException)
+        {
+          return !((RetriableRequestException) throwable).getDoNotRetryOverride();
+        }
+      }
+
+      return false;
     }
 
     /**
      * Retries a specific request.
      *
      * @param request Request to retry.
      * @param context Context of the retry request.
+     * @param numberOfRetryAttempts Number of retry attempts.
      * @return {@code true} if a request can be retried; {@code false} otherwise;
      */
-    public abstract boolean doRetryRequest(REQ request, RequestContext context);
+    public abstract boolean doRetryRequest(REQ request, RequestContext context, int numberOfRetryAttempts);
+  }
+
+  @ThreadSafe
+  private class ClientRetryTracker
+  {
+    private final int _aggregatedIntervalNum;
+    private final long _updateIntervalMs;
+    private final Clock _clock;
+    private final String _serviceName;
+
+    private final Object _counterLock = new Object();
+    private final Object _updateLock = new Object();
+
+    @GuardedBy("_updateLock")
+    private volatile long _lastRollOverTime;
+    private double _currentAggregatedRetryRatio;
```

Comment:
Minor: if the _updateLock is guarding it there should be a separate annotation for it

Reply:
Fixed